presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java

```diff
@@ -128,11 +131,41 @@ private RelationPlan createInsertPlan(Analysis analysis)
 
         TableMetadata tableMetadata = metadata.getTableMetadata(session, target);
 
+        List<String> visibleTableColumnNames = tableMetadata.getVisibleColumnNames();
+        List<ColumnMetadata> visibleTableColumns = tableMetadata.getVisibleColumns();
+
+        RelationPlan plan = createRelationPlan(analysis);
+
+        Optional<List<String>> insertColumns = analysis.getInsertColumns();
+        if (insertColumns.isPresent() && !insertColumns.get().equals(visibleTableColumnNames)) {
```

Comment:
You should be able to do this unconditionally. The optimizer should prune any redundant projections.